base/sort.jl

```diff
@@ -933,73 +932,172 @@ end
 3×3 Array{Int64,2}:
   9  -2  8
   7   3  5
  -1   6  4
-```
-"""
-function sortrows(A::AbstractMatrix; kws...)
-    rows = [view(A, i, :) for i in axes(A,1)]
-    p = sortperm(rows; kws...)
-    A[p,:]
-end
 
-"""
-    sortcols(A; alg::Algorithm=DEFAULT_UNSTABLE, lt=isless, by=identity, rev::Bool=false, order::Ordering=Forward)
-
-Sort the columns of matrix `A` lexicographically.
-See [`sort!`](@ref) for a description of possible
-keyword arguments.
-
-# Examples
-```jldoctest
-julia> sortcols([7 3 5; 6 -1 -4; 9 -2 8])
+julia> sortslices([7 3 5; 6 -1 -4; 9 -2 8], dims=2) # Sort columns
 3×3 Array{Int64,2}:
   3   5  7
  -1  -4  6
  -2   8  9
 
-julia> sortcols([7 3 5; 6 -1 -4; 9 -2 8], alg=InsertionSort, lt=(x,y)->isless(x[2],y[2]))
+julia> sortslices([7 3 5; 6 -1 -4; 9 -2 8], dims=2, alg=InsertionSort, lt=(x,y)->isless(x[2],y[2]))
 3×3 Array{Int64,2}:
   5   3  7
  -4  -1  6
   8  -2  9
 
-julia> sortcols([7 3 5; 6 -1 -4; 9 -2 8], rev=true)
+julia> sortslices([7 3 5; 6 -1 -4; 9 -2 8], dims=2, rev=true)
 3×3 Array{Int64,2}:
  7   5   3
  6  -4  -1
  9   8  -2
 ```
+
+# Higher dimensions
+
+`sortslices` extends naturally to higher dimensions. E.g., if `A` is a
+a 2x2x2 array, `sortslices(A, dims=2)` will slices along the 3rd dimension,
+passing the 2x2 slices in each dimension to the comparison function. Note that
+while there is no default order on higher-dimensional slices, you may use the
+`by` or `lt` keyword argument to specify such an order.
+
+If `dims` is a tuple, the order of the dimensions in `dims` is
+relevant and specifies the linear order of the slices. E.g., if `A` is three
+dimensional and `dims` is `(1,2)`, we will take slices in the dimension three,
+and then arrange the result back into the array in column-major order. If
+`dims` is `(2,1)` instead, the same slices will be taken, but the result order
+will be row-major instead.
+
+# Higher dimensional examples
+```
+julia> A = permutedims(reshape([4 3; 2 1; 'A' 'B'; 'C' 'D'], (2, 2, 2)), (1, 3, 2))
+2×2×2 Array{Any,3}:
+[:, :, 1] =
+ 4  3
+ 2  1
+
+[:, :, 2] =
+ 'A'  'B'
+ 'C'  'D'
+
+julia> sortslices(A, dims=(1,2))
+2×2×2 Array{Any,3}:
+[:, :, 1] =
+ 1  3
+ 2  4
+
+[:, :, 2] =
+ 'D'  'B'
+ 'C'  'A'
+
+julia> sortslices(A, dims=(2,1))
+2×2×2 Array{Any,3}:
+[:, :, 1] =
+ 1  2
+ 3  4
+
+[:, :, 2] =
+ 'D'  'C'
+ 'B'  'A'
+
+julia> sortslices(reshape([5; 4; 3; 2; 1], (1,1,5)), dims=3, by=x->x[1,1])
+1×1×5 Array{Int64,3}:
+[:, :, 1] =
+ 1
+
+[:, :, 2] =
+ 2
+
+[:, :, 3] =
+ 3
+
+[:, :, 4] =
+ 4
+
+[:, :, 5] =
+ 5
+```
 """
-function sortcols(A::AbstractMatrix; kws...)
-    cols = [view(A, :, i) for i in axes(A,2)]
-    p = sortperm(cols; kws...)
-    A[:,p]
+function sortslices(A::AbstractArray; dims, kws...)
+    if isa(dims, Symbol)
+        dims = dims == :rows ? 1 :
+               dims == :cols ? 2 :
+               error("Only :rows or :cols are allowed as symbolic dimensions")
+    end
+    _sortslices(A, Val{dims}(); kws...)
```

Comment:
Looks like this file is `include`d in the system image before `Val` is defined, which is causing the CI errors.

Reply:
I'm actually not entirely convinced that it should live here. I think it might make more sense to go with multidimensional, esp since we now use this file from the compiler. This code has very little to do with sorting.

Reply:
👍 That does seem like a more appropriate place for it, IMO.